### PR TITLE
dynamically add @resolver required fields to downstream query

### PR DESCRIPTION
**Pull request by CNAChino**
# What Changed
During schema build,  the field references (required fields) specified in field `@resolver`' directive are collected in FieldResolverContext.  And in the execution,  the required fields are check against the selected fields and adds them in the downstream query.

# Why
make the field resolver required fields optional in client query and have the graphql orchestrator automatically resolve those fields so that the required field values will be available for the field resolver.

Todo:
- [x] Add tests
- [x] Add docs

### CODE REVIEW

File: src/main/java/com/intuit/graphql/orchestrator/schema/transform/FieldResolverTransformerPostMerge.java [hunk 1]
```diff
@@ -53,6 +69,24 @@ public XtextGraph transform(XtextGraph sourceXtextGraph) {
     return sourceXtextGraph;
   }
 
+  private void validateRequiredFields(FieldResolverContext fieldResolverContext) {
```
Comment (kmoore-intuit) [inline src/main/java/com/intuit/graphql/orchestrator/schema/transform/FieldResolverTransformerPostMerge.java:72]: Possibly refactor and move the validation to its own class(FieldResolveContext) so the transformer performs one task. Also performs less peek operations making it cleaner

**Review by kmoore-intuit — APPROVED** [review]

File: src/main/java/com/intuit/graphql/orchestrator/schema/transform/FieldResolverTransformerPostMerge.java (hunk 1 above)
Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/schema/transform/FieldResolverTransformerPostMerge.java:72]: Yup agree.  Good suggestion.   Will refactor

Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/schema/transform/FieldResolverTransformerPostMerge.java]: this is done

File: src/main/java/com/intuit/graphql/orchestrator/utils/FieldReferenceUtil.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+   * @param inputString string to extract field references from
+   * @return set of field references
+   */
+  public static Set<String> getAllFieldReferenceFromString(String inputString) {
+    if (StringUtils.isEmpty(inputString)) {
+      return Collections.emptySet();
+    }
+
+    Set<String> output = new HashSet<>();
+    Pattern pattern = Pattern.compile("\\$\\w+"); // matches a-z A-Z _ 0-9
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/utils/FieldReferenceUtil.java:27]: keep this static

Comment (CNAChino) [reply src/main/java/com/intuit/graphql/orchestrator/utils/FieldReferenceUtil.java:27]: nice catch thanks.  This is done in next PR.  Though this is not used in execution.

**Pull request merged**